### ISSUE TITLE
Linux Mint - Can't find a way to install

### ISSUE THREAD

**Issue by sumnerevans** [sublime-music/sublime-music-python#216]
In GitLab by @SickWolf on May 23, 2020, 14:19

I am no expert and looked a lot around.

Trying to install using pip (fresh mint install):
```
pip2 install wheel setuptools
pip2 install sublime-music[keyring]
```

There are a bunch of `Failed building wheel for pycairo`

StackOverflow suggested to type `python setup.py bdist_wheel` before installing but it didn't seem to do anything. (I never understood python)

I also tried to make a flatpack as there are none in the release section, but it needs a bunch of sdk, some I cannot find (like gnome.Sdk).

I have no idea what I'm doing but I'd really like to use this with ND. And I'm sorry for opening an issue for this - I should be able to figure it out.

Thanks

**Comment by sumnerevans** [comment]
In GitLab by @robozman on May 23, 2020, 14:28

Sublime Music is written in Python 3, not 2. Try changing your `pip2`s to `pip3` and see if that helps.

**Comment by sumnerevans** [comment]
In GitLab by @robozman on May 23, 2020, 14:30

RE: flatpak ssiues
`org.gnome.Sdk` should be available in the flathub repo. Do you have flathub listed when you run `flatpak remotes`?

**Comment by sumnerevans** [comment]
In GitLab by @SickWolf on May 23, 2020, 15:02

I tried before with pip3 but I had pages of errors I've never seen before.
I now installed org.gnome.Sdk. Ran the script, error. installed org.gnome.Platform, reran the script, error. Installed cffi python module, etc.

Doing so though I noticed the same error popping up as when I try to install directly with python and not making a flatpak. (I mean, it makes sense) I double checked the folders and I noticed a few seconds ago the **flatpak-requirements.txt** file :/. I attached the log from `pip install -r flatpak-requirements.txt` command. (my pip uses python version 3.6)

I should've checked that sooner and avoid a bunch of trial and error, my bad.

[python-error.txt](https://gitlab.com/sublime-music/sublime-music/uploads/f7849535a4160b8389c404fa7964fa50/python-error.txt)

**Comment by sumnerevans** [comment]
In GitLab by @robozman on May 23, 2020, 15:34

A few things here:

Python version: I was taking to @sumnerevans and he said that Sublime Music uses some Python 3.8 specific features so you might be out of luck installing with pip without upgrading your Python version. 

`#include <Python.h>` error: This typically means you are missing the Python development headers from your system, which usually live in a package called something like `python3-dev` (I think that should be right for Mint).

`flatpak-requirements.txt`: You shouldn't be installing the dependencies in this file directly. This file is used by `flatpak-pip-generator` to automatically generate a `flatpak` `json` file that then will install the dependencies into the `flatpak` container when you run the `flatpak-builder` tool. 

@sumnerevans was saying that currently the flatpak build is not working, as the Freedesktop runtime `org.freedesktop.Sdk` does not provide Python 3.8 and Sublime requires that. So even if you sorted out all your issues you wouldn't actually be able to make a flatpak. If you would like to look into adding Python 3.8 to the flatpak container that would be awesome. I would be surprised if someone else on the internet isn't dealing with this issue as well. The solution might just be to build Python 3.8 from source in the flatpak build script. This is kinda gross, but I think that's what I remember the KDE runtime was doing as Python 2 is a build dependency for QtWebEngine, but maybe someone has come up with a sleeker solution for the non deprecated versions of Python :laughing:

In conclusion: If you want to try installing Sublime Music, I would say your best bet is finding a way to install Python 3.8 on your system. I'm not really sure what the best way to do that on Ubuntu derivatives (aka Mint) today. You might want to just Google around for the currently recommended solution. Last time I dealt with this I had to use a PPA, but I don't really use Ubuntu based distros much so I don't know if this is still the recommended method.

**Comment by sumnerevans** [comment]
In GitLab by @SickWolf on May 23, 2020, 15:55

Thank you @robozman that explains a lot!

Yeah packages tend to be quite outdated on non rolling distros (but I like the stability :/ ). I'll try to find a repo/tutorial to install Python 3.8 and update this issue if needed with maybe instructions for other potential users on Mint/Ubuntu - or switch distro.... maybe.

For the `flatpak-requirements.txt` well it's too late but it "only installed" some packages and won't affect me so no harm done.

On Ubuntu/Mint, there are a lot of python-*dev packages but I will fix the version issue before trying to figure this one out.

Thanks for the time @sumnerevans and you took to help with this. Your post is clear and easy to understand and I appreciate that. - I hate stuff that compiles, what a headache.

:heart: Stay safe.

Side note: I was using Amarok for 7+ years but for the past 2 years I've been using a locked in and limited music player (DS Audio and Synology Audio) I'm now trying to make my library more accessible and this project looks awesome.

**Comment by sumnerevans** [comment]
In GitLab by @SickWolf on May 23, 2020, 15:55

closed

**Comment by sumnerevans** [comment]
In GitLab by @robozman on May 23, 2020, 16:00

You're welcome! Hope you have luck getting it installed. 

AFAIK, pip unfortunately doesn't provide a easy way to uninstall the package tree installed by a requirements.txt file. I briefly saw some stack overflow bash magic that may work. OR you can just do what I always do and just `rm -rf` my `$HOME` python install directory.

**Comment by sumnerevans** [comment]
In GitLab by @SickWolf on May 23, 2020, 16:04

Oh I can do that? :P thanks for the tip!

For future reference as I tried it and it worked without any hiccups:
* [How to compile and install Python3.8 on Debian-based distro](https://tecadmin.net/install-python-3-8-ubuntu/)

```
# https://stackoverflow.com/questions/22100213/package-cairo-was-not-found-in-the-pkg-config-search-path-node-j-s-install-canv

sudo apt-get install libcairo2-dev libjpeg-dev libgif-dev
pip3.8 install pycairo

# https://pygobject.readthedocs.io/en/latest/getting_started.html
sudo apt install libgirepository1.0-dev gcc libcairo2-dev pkg-config python3-dev gir1.2-gtk-3.0
pip3.8 install pycairo
pip3.8 install PyGObject

pip3.8 install sublime-music[keyring]

$ sublime-music
AttributeError: /usr/lib/x86_64-linux-gnu/libmpv.so.1: undefined symbol: mpv_render_context_create

# I installed (but didn't work): https://non-gnu.uvt.nl/debian/bionic/mpv/mpv_0.32.0+wsl.2_amd64.deb

# https://askubuntu.com/questions/1128842/failed-building-minitube-with-compiled-libmpv-v0-29-0
# I can try and recompile libmpv, but I don't want to mess with too much packages and break a dependency
```

**Comment by sumnerevans** [comment]
Have you installed `libmpv-dev`?

**Comment by sumnerevans** [comment]
I believe @anarcat and @baldurmen also use Debian-based distros. Maybe one of could help here as well?

**Comment by sumnerevans** [comment]
In GitLab by @baldurmen on May 23, 2020, 18:43

I got a (mostly) working output using these packages:

* gir1.2-gtk-3.0
* gir1.2-nm-1.0
* python3-gi
* python3-mpv

I did have to package `python3-mpv`, as it wasn't in the archive already. Chances are it's not in Linux Mint's since it too recent (Debian Unstable). You can have a look at all the packages needed to build and run sublime in Debian [here](https://salsa.debian.org/python-team/applications/sublime-music/-/blob/debian/master/debian/control).

@sumnerevans sublime 0.9.1 is currently in [the Debian NEW queue](https://ftp-master.debian.org/new/sublime-music_0.9.1-1.html) and is awaiting approval from the team who reviews the first upload of a new package. It'll migrate in Debian unstable and Debian testing shortly after. :)

**Comment by sumnerevans** [comment]
@baldurmen Thank you for your work on that! It's very exciting.

Are there any stability expectations for packages in the Debian repos? The current plan is that the upgrade from 0.9.1 to 0.9.2 will require users to delete the cache and re-configure Sublime. (I think I will put some sort of banner notification if I detect old config informing the user of the breakage.) Once I stabilize with the 1.0 release, I will support config/cache migrations for all new versions.

**Comment by sumnerevans** [comment]
In GitLab by @baldurmen on May 24, 2020, 11:56

@sumnerevans no stability expectations for now. Packages in `unstable` can be buggy or unstable and we normally try to get a stable version ready for the next `stable` release, which should be around 2021-01-12.

But you don't really have to worry about that. When the freeze for the next `stable` release happens, I'll have a look and if the current packaged version in `unstable` isn't suitable for `stable`, I'll either revert to a previous, more stable version or ping you.
